### DISCUSSION TITLE
How to upload and get public url at once

### DISCUSSION THREAD

**Discussion by IHIutch** [supabase/supabase#1230] [Questions] [answered]
Asked this here, but figured I've give it its own question: https://github.com/supabase/supabase/discussions/1146#discussioncomment-640379

Looking for a pattern that uploads and gets a URL in the same process, anyone have a good example of how to do this?

Or is this example the suggested process, where you upload in one spot and use download/createSignedURL in another as needed?
https://github.com/supabase/supabase/blob/master/examples/nextjs-ts-user-management/components/Avatar.tsx

**Comment by inian** [comment, marked as answer]
Hi @IHIutch, 

You get the object key after you upload the object. You can store this key in your database and later fetch the object via [download](https://supabase.io/docs/reference/javascript/storage-from-download). 

Right now, objects in the buckets cannot be made public unless you explicitly call createSignedURL. We are working on a way to make an entire bucket public, so that you can access objects in the bucket without calling `createSignedURL`. The objects in those buckets can just be accessed with the object path and without any token.

**Reply by IHIutch** [reply]
Thanks!

**Reply by hjaber** [reply]
Now that public storage buckets are available, would it be possible to return .getPublicUrl() in the return object of .upload to reduce complexity?

```
const avatarFile = event.target.files[0]
const { data, error } = await supabase
  .storage
  .from('avatars')
  .upload('public/avatar1.png', avatarFile’})
```

`console.log(data)`
`{"Key":"avatars/public/avatar1.png", “PublicURL”: “https://acshf2sfdf.supabase.in/storage/v1/object/public/avatars/public/avatar1.png“`}`

**Reply by inian** [reply]
Hi @hjaber, `getPublicUrl` just does some string concatenation and doesn't make any extra network calls, so it shouldn't make things any slower if you make a separate call for it, if thats what you are worried about.

**Reply by hjaber** [reply]
I don’t feel strongly that the publicURL should be returned but I figured if the url is available after upload and the key is already being returned, why not include it in the return data object? It does add complexity to call two functions instead of one.

No need to follow up @inian, thanks for clarifying it doesn’t cause any performance issues, that’s good enough for me.

**Reply by calumk** [reply]
+1 - It is odd that the publicUrl isnt returned

**Reply by hjaber** [reply]
> +1 - It is odd that the publicUrl isnt returned

Still agree it should be returned but as @inian pointed out, you know the publicURL beforehand since it is just string concatenation and you don't actually need to call a function to get it. If the syntax changes in the future though, it would cause breaking changes. I think the syntax for the string is roughly as follows every time:

`https://projectRef.supabase.in/storage/v1/object/filePath/fileName.png`

**Reply by calumk** [reply]
Ahh I see, so its possible to simply concat the file name without another call - ok, that makes sense.

I think maybe the existance of the `getPublicUrl` function clouds this a little. - it makes it seem like it is _needed_ in order to query supabase for the public url.

Maybe documentation can be added to explain what `getPublicUrl` is actually doing, and that it is _optional_

**Reply by zaynelovecraft** [reply]
I have a CMS dashboard for admin to upload pictures to storage bucket. Can I generate signedURLs with a 10 year expiry time or forever (does it have to be set in seconds?). Then can I store that url in a table so I can query the table for all image urls. I am using next js. For instance a year is 31,536,000 seconds. so is this correct? - .createSignedUrl('folder/avatar1.png', 31,536,000) I need the urls to last until deleted. I get urls with Next js Get static props ISR. Please help Thank you!

**Reply by hypervillain** [reply]
@EpicDevv this might help https://stackoverflow.com/a/25201898
